### PR TITLE
Add option to build generated emulation code in the build directory.

### PR DESCRIPTION
**Pull request by dnovillo**
This patch adds support for building the generated emu headers in
the build directory instead of the current scheme.  We need it
for our internal build system, and to avoid polluting the source
directory with generated code.

In addition, it speeds up the generation of header files by removing
the string concatenation done during the conversion to hex strings.
In python 2, that can take hours to complete (we still use python 2
in our build system by default).

The patch re-factors the generation of the header files to avoid
repeating common code.

I have tested this by comparing the headers generated.  The results
are identical.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

File: patch/generate/script/genGlslEmuLib.py [hunk 1]
```diff
@@ -33,37 +33,61 @@
...
-        os.chdir("./" + gfx)
-
-        for f in os.listdir("./"):
-            if f.startswith("g_") or  f.endswith(".bc") or f.endswith(".lib"):
-                print(">>>  (LL-clean) remove " + gfx + "/" + f)
-                os.remove(f)
-
-        os.chdir("./..")
+        for subf in os.listdir(fpath):
+            subfpath = os.path.join(fpath, subf)
```
Comment (amdrexu) [inline patch/generate/script/genGlslEmuLib.py]: subfpath -> subfPath

File: patch/generate/script/genGlslEmuLib.py [hunk 2]
```diff
@@ -223,24 +219,31 @@
...
-    for f in os.listdir(gfx):
-        if f.endswith(".ll"):
-            cmd = LLVM_AS + " " + gfx + "/" + f
-            print(">>>  (LL-as) " + cmd)
-            if OS_TYPE == "win" :
-                subprocess.check_call(cmd)
-            else :
-                subprocess.check_call(cmd, shell = True)
+    for src_dir in SRC_DIRS:
+      gfx_subdir = os.path.join(src_dir, gfx)
```
Comment (amdrexu) [inline patch/generate/script/genGlslEmuLib.py]: gfx_subdir -> gfxSubdir

**Comment by dnovillo** [comment]
Thanks for the prompt review.  I've addressed your comments.  PTAL.

**Comment by JacobHeAMD** [comment]
ok to test

**Comment by JacobHeAMD** [comment]
test this please

**Comment by dnovillo** [comment]
@JacobHeAMD I found no tests for this file, so I could only test it by comparing the generated header files before and after my patch.  They were identical.

**Comment by dnovillo** [comment]
@JacobHeAMD Additionally, I see that there is a test status stating that there is 1 failing check.  But there is no link to the test.  Where can I find it?

**Comment by JacobHeAMD** [comment]
Hi @dnovillo , sorry that there is nowhere to get the build log for you so far because it's not an issue of your patch. There is an issue with test framework and I'll retrigger the test once it's fixed.

**Comment by jinjianrong** [comment]
@dnovillo there are still some problems with the automated build and test farm, we are working on it. please ignore the build and test status for now. thanks.

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by dnovillo** [comment]
> retest this please

@JacobHeAMD is this directed at me?  I'm not sure.  I've already stated that I don't know how to test this, other than doing the test I've already done.  If you point me at instructions on how to test this, I'll do it.

**Comment by JacobHeAMD** [comment]
Hi @dnovillo , please note that "r e t e s t   t h i s  p l e a s e" "t e s t   t h i s  p l e a s e" is used to trigger the test.

**Comment by JacobHeAMD** [comment]
@dnovillo  Sorry for the confusion. :)

**Comment by dnovillo** [comment]
Oh!  Ha ha.  OK, thanks :)

**Comment by amdvlk-admin** [comment]
Hi @dnovillo , sorry for delay. Could you please help to rebase it? And merge it into one commit is better.

**Comment by linqunzju** [comment]
retest this please

**Comment by JacobHeAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4c1fc3cf65d393ae1183cba55646a7db9c0b5978
<details open><summary>Driver branches used in build</summary><ul>

<li>PAL: dev d939792</li>
<li>SPVGEN: dev 328a099</li>
<li>XGL: dev 4b820f0</li>
<li>LLVM: amd-vulkan-dev e594994ad72</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 795/795 (100.0%)</li>
<li>Failed: 0/795 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>ubuntu, 18-04, ellesmere</summary><ul>

<li>Passed: 312610/312610 (100.0%)</li>
<li>Failed: 0/312610 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JacobHeAMD** [comment]
Hi @amdrexu , could you please review this pull request?

**Comment by JacobHeAMD** [comment]
The pull request is outdated already and there is no respond from author. Let's close it. It will be ported to internal branch manually. And the option will be supported soon.

**Comment by dnovillo** [comment]
Sorry about the delay. I was out for the last 2 weeks. Anything for me to
do? Should I send a new PR?

**Comment by JacobHeAMD** [comment]
Thanks, @dnovillo . I'm merging your change into internal branch (almost done). So it's unnecessary to send a new PR if you don't mind there is no your commit in the git log.  :)

**Comment by dnovillo** [comment]
On Tue, Dec 4, 2018 at 9:08 AM Jacob He <notifications@github.com> wrote:

> Thanks, @dnovillo <https://github.com/dnovillo> . I'm merging your change
> into internal branch (almost done). So it's unnecessary to send a new PR if
> you don't mind there is no your commit in the git log. :)
>
Great! Thanks.  No worries about the git log :)

>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4c1fc3cf65d393ae1183cba55646a7db9c0b5978
<details open><summary>Driver branches used in build</summary><ul>

<li>PAL: dev d939792</li>
<li>SPVGEN: dev 328a099</li>
<li>XGL: dev 4b820f0</li>
<li>LLVM: amd-vulkan-dev e594994ad72</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 795/795 (100.0%)</li>
<li>Failed: 0/795 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>ubuntu, 18-04, ellesmere</summary><ul>

<li>Passed: 312610/312610 (100.0%)</li>
<li>Failed: 0/312610 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request closed without merging**